src/main/java/io/aiven/kafka/auth/VerdictCache.java

```diff
@@ -16,44 +16,63 @@
 
 package io.aiven.kafka.auth;
 
+import javax.annotation.Nonnull;
+
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import org.apache.kafka.common.security.auth.KafkaPrincipal;
 
+import io.aiven.kafka.auth.json.AclPermissionType;
 import io.aiven.kafka.auth.json.AivenAcl;
 
 public class VerdictCache {
-    private final List<AivenAcl> aclEntries;
+    private final List<AivenAcl> allowAclEntries;
+    private final List<AivenAcl> denyAclEntries;
     private final Map<String, Boolean> cache = new ConcurrentHashMap<>();
 
-    private VerdictCache(final List<AivenAcl> aclEntries) {
-        this.aclEntries = aclEntries;
+    private VerdictCache(@Nonnull final List<AivenAcl> denyAclEntries, @Nonnull final List<AivenAcl> allowAclEntries) {
+        this.denyAclEntries = denyAclEntries;
+        this.allowAclEntries = allowAclEntries;
     }
 
     public boolean get(final KafkaPrincipal principal,
                        final String operation,
                        final String resource) {
-        if (aclEntries != null) {
-            final String cacheKey = resource
-                + "|" + operation
-                + "|" + principal.getName()
-                + "|" + principal.getPrincipalType();
-            final Predicate<AivenAcl> matcher = aclEntry ->
-                aclEntry.check(principal.getPrincipalType(), principal.getName(), operation, resource);
-            return cache.computeIfAbsent(cacheKey, key -> aclEntries.stream().anyMatch(matcher));
-        } else {
-            return false;
-        }
+        final String principalType = principal.getPrincipalType();
+        final String cacheKey = resource
+            + "|" + operation
+            + "|" + principal.getName()
+            + "|" + principalType;
+
+        final Predicate<AivenAcl> matcher = allowAcl ->
+            allowAcl.match(principalType, principal.getName(), operation, resource);
```

Comment:
Nit:
```suggestion
        final Predicate<AivenAcl> matcher = acl ->
            acl.match(principalType, principal.getName(), operation, resource);
```
or something of this sort ('cause it's used for allow and deny as well).